Feature/rviz media export

Martin-Idel-SI opened the pull request:
This PR resolves #4 in the following way:

- Resources can be added as rviz_ogre_media_exports to the ament_index
- For convenience, a simple cmake macro taking the directories to be added as resources is provided
- RViz queries the ament_index at runtime adding all resources it finds
- The user is informed, if some of the claimed resource folders are missing

Martin-Idel-SI committed:
Implement replacement for rviz' media_export

- resources can be an arbitrary number of shallow folders
- should be registered as rviz_ogre_media_exports resource in ament_index
- a warning is thrown if a non-existant folder is added as a resource location

Martin-Idel-SI committed:
Introduce a macro to simplify adding resources

- macro takes an arbitrary number of paths to folders
- checks if folders exists
- registers all resources in ament_index
- installs all folders

Martin-Idel-SI committed:
Extract function to unit test and add tests

wjwwood commented:
Sorry, I didn't think to do this one first, but the style pr I just merged means this needs to be rebased now.

wjwwood commented on rviz_rendering/CMakeLists.txt:
```diff
@@ -47,7 +48,7 @@ endif()
 set(headers_with_q_object
   include/rviz_rendering/render_window.hpp
-)
+  src/rviz_rendering/string_helper.hpp)
```
Leave the `)` on the new line, the whole point is to simplify the diff :)

wjwwood commented on rviz_rendering/src/cmake/register_rviz_ogre_media_exports.cmake:
```diff
@@ -0,0 +1,56 @@
+        set(RESOURCE_FILE_CONTENT "${RESOURCE_FILE_CONTENT}${PROJECT_NAME}/${DIR}\n")
+    endforeach()
+
+    ament_index_register_resource(rviz_ogre_media_exports CONTENT ${RESOURCE_FILE_CONTENT})
```
Because this is written within this function, you should make it clear that this function can only be called once in the doc string.

A different way to do this would be to have each call to this function build up the content and then have a hook that runs in `ament_package()` which actually writes the resource.

See the other `ament_*` functions for examples:

- adds to variable in each call: https://github.com/ament/ament_cmake/blob/master/ament_cmake_core/cmake/environment_hooks/ament_environment_hooks.cmake#L78
- triggered via a hook: https://github.com/ament/ament_cmake/blob/36d5c721f5d9b6242126175e9002990017f849b0/ament_cmake_core/ament_cmake_environment_hooks-extras.cmake#L25-L26
- and consumed in macro called from hook file: https://github.com/ament/ament_cmake/blob/master/ament_cmake_core/cmake/environment_hooks/ament_generate_package_environment.cmake#L54

Martin-Idel-SI replied:
If your other proposition is the "usual" way of doing things for ament, then I'd prefer to also go this way to avoid future confusion. I'll have a look at it.

wjwwood commented on rviz_rendering/src/rviz_rendering/render_system.hpp:
```diff
@@ -52,6 +49,11 @@
+#include <QDir>
+
+#include <cstdint>
+#include <string>
```
The system include should happen first, I know that makes cpplint unhappy, but that's the right order to do them in. I just disable the check on the Qt includes (on which it complains). This is the wrong order in my opinion, based on the list here:

https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

Martin-Idel-SI commented:
@wjwwood No problem. I did the rebase.

wjwwood commented:
Sorry forgot to post my comments from yesterday until just now :/

Martin-Idel-SI committed:
Change CMake macro

- macro now only saves to variable and can be added
an arbitrary number of times
- the true installation is done with ament_package()
- furthermore, we correctly deploy the macro

Martin-Idel-SI commented:
I implemented a hook similar to the files you pointed out. 

Now, if I have a simple ros2 package, I can provide folders to rviz_ogre_media at will and they will be collected and installed at the end. For instance a call like this: 

`find_package(rviz_rendering REQUIRED)`
`register_rviz_ogre_media_exports(DIRECTORIES "ogre_media")`
`register_rviz_ogre_media_exports(DIRECTORIES "second_media")`
`register_rviz_ogre_media_exports(DIRECTORIES "second_media")`

would install the folders `ogre_media` and `second_media` (the latter twice, but the second time, it's clearly up-to-date) with `ament_package()`. Please review again, thank you!

wjwwood commented on rviz_rendering/src/rviz_rendering/render_system.hpp:
```diff
@@ -52,6 +49,11 @@
+#include <cstdint>
+#include <string>
+
+#include <QDir> // NOLINT
```
These should be included before the Ogre headers and outside of the warnings pragma. I would just move them back...

wjwwood commented on rviz_rendering/src/rviz_rendering/string_helper.hpp:
```diff
@@ -0,0 +1,49 @@
+{
+public:
+  static std::vector<std::string>
+  splitStringIntoTrimmedItems(const std::string &, const char);
```
Why does this need to be a static method of a class? Would it not be just as good as a free function in a namespace?

It seems odd to be able to instantiate a `string_helper` class which can do nothing special.

wjwwood reviewed (commented):
It would be great to have a test for this feature, though I know it would require another package. I could imagine having a package embedded within this one which is built as part of the tests, but it might be just as well to have an example package which demonstrates how this works as well as tests it.

Martin-Idel-SI committed:
Implement review comments

Fix StringHelper as its own class

- Will never get instantiated
- Now only provides a method to the rviz_rendering namespace

Fix include order

Martin-Idel-SI committed:
Add example plugin to test register_ogre_media_exports

- the plugin could be expanded to a true rviz_plugin
- this can be used as an example for plugin developers

Martin-Idel-SI commented:
Yes, I was also thinking about the testing issue. 

I had a small example set up locally, so I'll clean this up and add automated tests of the functionality. Since at some point, we need a developer documentation, we will probably want to have a rather minimal example plugin anyway, so maybe it's a good idea to start setting this up right now. We could then evolve the example into a true rviz example plugin in due course.

Martin-Idel-SI committed:
Make CMake output more meaningful by adding full path

Martin-Idel-SI commented:
I updated the PR with requested changes and new tests testing the macro (all tests are passing on Linux at least). I decided on an extra package, but of course, we can change this (especially the name).

wjwwood commented on rviz_rendering/CMakeLists.txt:
```diff
@@ -126,6 +129,12 @@ if(TARGET Qt5::windeployqt)
+include(${CMAKE_CURRENT_SOURCE_DIR}/src/cmake/register_rviz_ogre_media_exports_hook-extras.cmake)
+
+install(DIRECTORY src/cmake
+        DESTINATION share/${PROJECT_NAME})
```
Please don't do aligned indentations (c++ style check will catch this, but none for CMake). Instead do:

```cmake
install(DIRECTORY src/cmake
  DESTINATION share/${PROJECT_NAME})
```

Or:

```cmake
install(
  DIRECTORY src/cmake
  DESTINATION share/${PROJECT_NAME})
```

Please check the rest of the files, I'm seeing it a lot.

wjwwood commented on rviz_rendering/CMakeLists.txt:
```diff
@@ -142,6 +151,16 @@ if(BUILD_TESTING)
+        "${CMAKE_CURRENT_SOURCE_DIR}/src/cmake/register_rviz_ogre_media_exports_hook-extras.cmake"
+        "${CMAKE_CURRENT_SOURCE_DIR}/src/cmake/register_rviz_ogre_media_exports.cmake")
+
+ament_package()
```
Please make sure all the files have a newline at the end of the file (again the C++ style checker would catch this). You can see it on github like this:

<img width="170" alt="screen shot 2017-10-25 at 8 24 46 pm" src="https://user-images.githubusercontent.com/100427/32033651-90d74132-b9c2-11e7-8220-1bbe187ac360.png">

wjwwood commented on rviz_rendering/src/cmake/register_rviz_ogre_media_exports_hook-extras.cmake:
```diff
@@ -0,0 +1,32 @@
+# POSSIBILITY OF SUCH DAMAGE.
+
+option(AMENT_RVIZ_OGRE_MEDIA
+        "Add ogre_media resources to RViz" ON)
```
This is neither aligned indention nor the standard two-space indention. I see this a lot in the file too.

wjwwood commented on rviz_rendering/src/rviz_rendering/render_system.cpp:
```diff
@@ -39,21 +40,25 @@
 #include <GL/glx.h>
 #endif
 
+
```
Please avoid unnecessary whitespace changes and minimize vertical whitespace: https://google.github.io/styleguide/cppguide.html#Vertical_Whitespace

wjwwood commented on rviz_rendering/src/rviz_rendering/render_system.cpp:
```diff
@@ -330,32 +335,34 @@ RenderSystem::setupResources()
+    {
+      std::vector<std::string> filenames =
+        string_helper::splitStringIntoTrimmedItems(
+        content, '\n');
```
nitpick: I would restyle this to be:

```c++
      std::vector<std::string> filenames =
        string_helper::splitStringIntoTrimmedItems(content, '\n');
```

Or:

```c++
      std::vector<std::string> filenames = string_helper::splitStringIntoTrimmedItems(
        content, '\n');
```

But not both. You don't need to fix this particular case, but just for future reference.

wjwwood commented on rviz_rendering/src/rviz_rendering/render_system.hpp:
```diff
@@ -56,6 +56,8 @@
 # pragma GCC diagnostic pop
 #endif
 
+#include <QDir> // NOLINT
```
Two spaces between comment and code please:

```c++
#include <QDir>  // NOLINT
```

wjwwood commented on rviz_rendering_tests/package.xml:
```diff
@@ -0,0 +1,30 @@
+    <license>BSD</license>
+
+    <!-- TODO(Martin-Idel-SI): website does not exist -->
+    <url type="website">http://ros.org/wiki/rviz_rendering_tests</url>
```
Just leave this out for now rather than TODO it.

wjwwood reviewed (changes requested):
I had a few style comments, which btw I would have taken care of for you, but I don't seem to have push permission on this pr for some reason. (note I don't need it on the repository, just the branch and the pr normally does this for you unless you opt out when creating it)

I'm ok with the new package right now. Later we might roll it into a rviz plugin example package where we can just as easily do the same tests.

Martin-Idel-SI committed:
Adjust CMake style to correct style

Martin-Idel-SI committed:
Adjust style in test plugin

Martin-Idel-SI committed:
Adjust code formatting

Martin-Idel-SI commented:
Hopefully fixed all code-style issues. It's good to bring them up - I'm relying on my IDE + linting to handle the code style and apparently I still had some problems with the configuration. 

By the way, in order to simplify reviewing, I'm not squashing. If you want me to squash beforehand, I can do that.

Martin-Idel-SI committed:
Remove anonymous namespaces around tests

- Unnecessary
- Multiple Tests inside one namespace fail cppcheck >=1.79

Martin-Idel-SI commented:
I also had to remove the anonymous namespace, since it makes cppcheck >1.78 fail for some unknown reason.

Martin-Idel-SI commented:
I rebased everything onto the current master and hopefully all requested changes are implemented.

wjwwood commented on rviz_rendering/CMakeLists.txt:
```diff
@@ -107,41 +110,57 @@ target_link_libraries(rendering_example rviz_rendering Qt5::Widgets)
-        COMMAND
-          Qt5::windeployqt
-          --dir "${CMAKE_CURRENT_BINARY_DIR}/windeployqt"
-          "$<TARGET_FILE_DIR:rendering_example>/$<TARGET_FILE_NAME:rendering_example>"
```
Re-indenting to two space is correct, but the extra indentation for the command's arguments should have stayed. You don't have to fix this, but for the future reference.

wjwwood commented on rviz_rendering/src/rviz_rendering/string_helper.cpp:
```diff
@@ -0,0 +1,59 @@
+        });
+    auto whitespace_back = std::find_if_not(item.rbegin(), item.rend(), [](char character) {
+          return std::isspace<char>(character, std::locale(""));
+        });
```
The indentation doesn't look right here. I'll check it tomorrow morning, but make sure you have the latest ros2 sources I think this was fixed recently in our fork of uncrustify.

Martin-Idel-SI replied:
I just updated my ros2 installation but uncrustify doesn't seem to complain. I agree that it's not the best, though.

wjwwood replied:
Hmm, that sucks. Maybe not all the cases involving lambda's is fixed. Oh well, better to have it consistent with the tool, but it's always frustrating when the linter makes us do something we don't want 💢.

wjwwood commented:
I'll redo the uncrustify later.
